File: day-11/python/main.py
import sys
import copy
import math
from collections import deque


class Monkey(object):
    def __init__(self):
        self.items = None
        self.inspect = None
        self.throw = None
        self.activity = 0
        self.div = None


def parse_monkey(fd):
    m = Monkey()
    tst = None
    true = None
    false = None
    for line in fd:
        l = line.strip().replace(',', '')
        if l.startswith("Starting"):
            s = l.split()
            m.items = deque([int(t) for t in s[2:]])
        elif l.startswith("Operation"):
            f = l.split('=')[1]
            m.inspect = lambda old: eval(f)
        elif l.startswith("Test"):
            tst = int(l.split()[3])
            m.div = tst
        elif l.startswith("If true"):
            true = int(l.split()[5])
        elif l.startswith("If false"):
            false = int(l.split()[5])
        else:
            assert len(l) == 0
            break
    m.throw = lambda x: eval(f"{true} if (x % {tst} == 0) else {false}")
    return m


def parse_input():
    if len(sys.argv) != 2:
        sys.exit('I need a file')
    data = []
    with open(sys.argv[1]) as fd:
        for line in fd:
            if line.startswith("Monkey"):
                data.append(parse_monkey(fd))
            else:
                assert(False)
    return data


def common(monkeys, rounds, relax):
    for r in range(rounds):
        for m in monkeys:
            while True:
                try:
                    el = m.items.popleft()
                except:
                    break
                m.activity += 1
                el = m.inspect(el)
                el = relax(el)
                monkeys[m.throw(el)].items.append(el)
    activities = sorted([m.activity for m in monkeys])
    return activities[-2] * activities[-1]


def problem_1(monkeys):
    return common(monkeys, 20, lambda x: x // 3)


def problem_2(monkeys):
    # find "relax" ratio
    denom = math.prod([t.div for t in monkeys])
    return common(monkeys, 10000, lambda x: x % denom)


def main():
    data = parse_input()
    print("Day X, problem 1: ", problem_1(copy.deepcopy(data)))
    print("Day X, problem 2: ", problem_2(data))


if __name__ == '__main__':
    main()

File: day-11/rust/src/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;
use std::collections::HashSet;

mod monkey;
use monkey::Monkey;

fn parse_args() -> String {
    let mut args = env::args();
    if args.len() != 2 {
        eprintln!("Invalid arguments: I need the input file and nothing else");
        process::exit(1);
    }
    args.nth(1).expect("Already checked")
}

fn create_reader(filename: &String) -> impl Iterator<Item = String> {
    let file = File::open(&filename).unwrap_or_else(|_| {
        eprintln!("Cannot open file: {}", &filename);
        process::exit(1);
    });
    io::BufReader::new(file)
        .lines()
        .map(|l| l.unwrap())
}


fn parse_input(filename: &String) -> Vec<Monkey> {
    let mut lines = create_reader(&filename);
    let mut data = Vec::<Monkey>::new();
    loop {
        data.push(Monkey::parse(&mut lines));
        match lines.next() {
            Some(s) => {
                assert!(s.trim().len() == 0);
            },
            None => { break; }
        }
    }
    data
}

fn simulate(monkeys: &mut Vec<Monkey>,
            rounds: u32,
            relax: impl Fn(u64) -> u64) -> u64 {
    for _ in 0..rounds {
        for i in 0..monkeys.len() {
            let mut m = &mut monkeys[i];
            while let Some(mut el) = m.items.pop_front() {
                m.activity += 1;
                el = m.inspect(el);
                el = relax(el);
                let target_index = m.forward(el);
                m = &mut monkeys[target_index];
                m.items.push_back(el);
                m = &mut monkeys[i];
            }
        }
    }
    monkeys.sort_by(|a, b| a.activity.cmp(&b.activity));
    monkeys.iter().rev().take(2).map(|x| x.activity as u64).product()
}

fn problem_2(monkeys: &mut Vec<Monkey>) -> u64 {
    let divs: HashSet<u64> = monkeys.iter().map(|x| x.divisible_by).collect();
    let relax_factor: u64 = divs.iter().product();
    simulate(&mut monkeys.clone(), 10000, |x| x  % relax_factor)
}

fn main() {
    let filename = parse_args();
    let monkeys = parse_input(&filename);

    println!("Day 11, problem 1: {}",
             simulate(&mut monkeys.clone(), 20, |x| x / 3));
    println!("Day 11, problem 2: {}", problem_2(&mut monkeys.clone()));
}
